PeDuCoML/lib/parser.ml

```diff
@@ -100,28 +58,131 @@ let parse_capitalized_entity =
   else fail "Parsing error: not a capitalized entity."
 ;;
 
-let data_constructors = [ "Ok"; "Error"; "Some"; "None" ]
-
 let keywords =
-  [ "let"
-  ; "rec"
-  ; "match"
-  ; "with"
-  ; "if"
-  ; "then"
-  ; "else"
-  ; "in"
-  ; "fun"
-  ; "and"
-  ; "effect"
-  ; "type"
-  ; "perform"
-  ; "continue"
-  ]
+  [ "let"; "rec"; "match"; "with"; "if"; "then"; "else"; "in"; "fun"; "and"; "type" ]
 ;;
 
 (* ------- *)
 
+(* Patterns parsers *)
+
+let parse_pattern d =
+  choice
+    [ d.parse_ptuple d
+    ; d.parse_pconstruct_list d
+    ; d.parse_plist d
+    ; d.parse_pliteral
+    ; d.parse_pidentifier
+    ; d.parse_wild
+    ]
+;;
+
+let parse_pliteral =
+  fix
+  @@ fun self ->
+  remove_spaces
+  *> (parens self
+      <|>
+      let is_digit = function
+        | '0' .. '9' -> true
+        | _ -> false
+      in
+      let parse_int_literal =
+        take_while1 is_digit >>| int_of_string >>| fun x -> LInt x
+      in
+      let parse_string_literal =
+        char '"' *> take_while (( != ) '"') <* char '"' >>| fun x -> LString x
+      in
+      let parse_char_literal = char '\'' *> any_char <* char '\'' >>| fun x -> LChar x in
+      let parse_bool_literal =
+        string "true" <|> string "false" >>| bool_of_string >>| fun x -> LBool x
+      in
+      let parse_unit_literal = string "()" >>| fun _ -> LUnit in
+      let parse_literal =
+        choice
+          [ parse_int_literal
+          ; parse_string_literal
+          ; parse_char_literal
+          ; parse_bool_literal
+          ; parse_unit_literal
+          ]
+      in
+      pliteral <$> parse_literal)
+;;
+
+let parse_pidentifier =
+  fix
+  @@ fun _ ->
+  remove_spaces
+  *>
+  let parse_identifier =
+    parse_uncapitalized_entity
+    >>= fun entity ->
+    if List.exists (( = ) entity) keywords
+    then fail "Parsing error: keyword used."
+    else if String.for_all (fun c -> c = '_') entity
+    then fail "Parsing error: this is a wildcard."
+    else return @@ pidentifier entity
+  in
+  parse_identifier
+;;
+
+let parse_wild = remove_spaces *> (pwild <$> take_while1 (fun c -> c = '_'))
```

Comment:
Wildcard can be framed in brackets. This function cannot parse it. Fix, please

Reply:
Done